src/ui/hooks/FlowRouter.tsx

```diff
@@ -0,0 +1,205 @@
+    });
+
+    const prevStep: string = flow.current[state.step]![FlowDispatchTypes.BACK] ?? '';
+
+    let TheComponent = renderScreen(state.step),
+        componentStoreMethods: ComponentStoreMethods = createComponentStoreConnection(state.step);
+
+    useEffect(() => {
+        flow.current = createFlow(state.authIndex);
+    }, [state.authIndex]);
+
+    useEffect(() => {
+        // eslint-disable-next-line
+        componentStoreMethods = createComponentStoreConnection(state.step);
+    }, [state.step, createComponentStoreConnection]);
+
+    function flowReducer(state: any, action: FlowAction): any {
+        const {type, payload} = action;
+        const {step} = state;
+        const theFlow: Flow = flow.current;
+
+        if (!theFlow.hasOwnProperty(step)) throw new Error(`Referenced step '${step}' does not exist in the flow`);
+
+        switch (type) {
+        case FlowDispatchTypes.NEXT:
+            return {
+                ...state,
+                step: theFlow[step]![FlowDispatchTypes.NEXT]
+            };
+        case FlowDispatchTypes.BACK:
+            return {
+                ...state,
+                step: theFlow[step]![FlowDispatchTypes.BACK]
+            };
+        case FlowDispatchTypes.RESTART:
+            return {
+                ...state,
+                step: 'confirmation'
+            };
+        case FlowDispatchTypes.SET_AUTH_METHOD:
+            if ('menu' === state.step) {
+                return {
+                    ...state,
+                    authIndex: payload
+                };
+            } else {
+                throw new Error('Please don\'t try to set the authentication method from outside the Authentication Menu screen');
+            }
+        default:
+            throw new Error('Unknown action type');
+        }
+    }
+
+    function renderScreen(step: string) {
+        const componentMap: ComponentMap = {
+            confirmation: 'ConfirmationScreen',
+            menu: 'AuthenticationOptionMenu',
+            details: 'ResultDetails',
+            agency_qr: 'AgencyQR',
+            email_input: 'EmailScreen',
+            smsotp: 'SMSOTPScreen',
+            verificationRequirement: 'VerificationRequirementScreen'
+        };
+
+        const component: any = React.lazy(() => import('../screens/' + componentMap[step]));
+
+        return component;
+    }
+
+    // TODO: Make into a hook and inject the component store as a parameter
+    function createComponentStoreConnection(step: string) {
+        const get: ComponentStoreGet = (dataKey: string, dfault?: any, component?: string) => {
+            // This is equivalent to component ??= step, but that broke the compiler
+            component ?? (component = step);
+            dfault ?? (dfault = undefined);
+
+            return (componentStore.current[component] && componentStore.current[component][dataKey]) ?? dfault;
+        };
+
+        const set: ComponentStoreSet = (dataKey: string, value: any) => {
```

Comment:
I think `createComponentStoreConnection` should probably be a hook, but I don't necessarily see why the definitions of `get` and `set` should be a hook if we're not going to share these functions (methods?) to other components. The ability to even define them should be the sole responsibility of either the `FlowController` or some other hook that the `FlowController` invokes.

Reply:
Ok, turns out that a hook will not work here if we accept that the methods should change every time `state.step` is updated (I would argue in favor of retaining that behavior, because that will prevent collisions in `set`), we are breaking the Rules of Hooks if we try invoking a hook within `useEffect` https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level